### PR TITLE
Geomap: Implement Heatmap overlays

### PR DESCRIPTION
**Pull request by euniceek**
**What this PR does / why we need it**:
This PR implements the Heatmap overlays for the GeoMap plugin using [OL Heatmap layer](https://openlayers.org/en/latest/apidoc/module-ol_layer_Heatmap-Heatmap.html), addressing part of Worldmap parity #36585. 
The logic is similar to circle overlays PR #36680 where it calls the `dataFrameToLocations` function to receive a data object with fields such as values, lat, lng, and then maps each data value into a new `Point` which gets added to the Heatmap layer source. Users are given the options to choose the size and blur of each cluster. 

Note that on data update, the previous feature set, which includes the previous data points, have to first be cleared. This is to prevent drawing new points on top of the previous heatmap points. 

`normalize` function normalizes the data values to input for weight function in the vector layer. The reason for this is because heatmap weights only accept normalized value (0 ~ 1). However, any values below 0.1 is barely visible on the world map meaning small data values will not be visualized clearly. To prevent this, the function normalizes and then scales such that it returns a value in a range of range of 0.1 ~ 1.

image

image

**Special notes for your reviewer**:
`getColors` return undefined for color schemes options such as threshold and single color. These options don't make much sense in terms of gradient for OpenLayers (array of colors, weights are handled separately), it is currently being defaulted to the default gradient of heatmap. 

cc @alolita

### CODE REVIEW

File: public/app/plugins/panel/geomap/layers/data/heatMap.tsx [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+          // Set the gradient back to default if threshold or single color is chosen
+          vectorLayer.setGradient(['#00f', '#0ff', '#0f0', '#ff0', '#f00']);
+        }
+      },
+    };
+  },
+  // Heatmap overlay options
+  registerOptionsUI: (builder) => {
+    builder
+      .addSelect({
```
Comment (ryantxu) [inline public/app/plugins/panel/geomap/layers/data/heatMap.tsx:126]: With https://github.com/grafana/grafana/pull/36768, you can now remove all the queryFormat+spatial type configs

File: public/app/plugins/panel/geomap/layers/data/heatMap.tsx [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+        path: 'fieldMapping.metricField',
+        name: 'Metric Field',
+        defaultValue: defaultOptions.fieldMapping.metricField,
+        settings: {
+          filter: (f) => f.type === FieldType.number,
+          noFieldsMessage: 'No numeric fields found',
+        },
+      })
+      .addSliderInput({
+        path: 'radius',
```
Comment (ryantxu) [inline public/app/plugins/panel/geomap/layers/data/heatMap.tsx:204]:
```suggestion
        path: 'config.radius',
```
Side effect from: https://github.com/grafana/grafana/pull/36768

File: public/app/plugins/panel/geomap/layers/data/heatMap.tsx (hunk 1 above)
Comment (eunice98k) [reply public/app/plugins/panel/geomap/layers/data/heatMap.tsx]: Updated!

**Review by ryantxu — APPROVED** [review]
Simple and works well!  
image

File: public/app/plugins/panel/geomap/layers/data/heatMap.tsx [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+        // Get data points (latitude and longitude coordinates)
+        const info = dataFrameToPoints(frame, matchers);
+        if(info.warning) {
+          console.log( 'WARN', info.warning);
+          return; // ???
+        }
+
+        // Get the field of data values
+        const field = frame.fields.find(field => field.type === FieldType.number); // TODO!!!!
```
Comment (ryantxu) [inline public/app/plugins/panel/geomap/layers/data/heatMap.tsx:81]: We should make this configurable, and default to "count" -- for configuration, lets use the work starting in:
https://github.com/grafana/grafana/pull/36825

so I will go ahead and merge this.

Great work @eunice98k

**Pull request merged**